### ISSUE TITLE
Skeletonize large volumes

### ISSUE THREAD

**Issue by unidesigner** [seung-lab/kimimaro#8]
Hi, very nice library!

Out-of-curiosity, how do you use it to go about skeletonization of a petascale image volume? I'd think that stiching of extracted skeletonizations from (adjacent) subvolumes/blocks will cause some artefacts.

Thanks, Stephan

**Comment by william-silversmith** [comment]
Hi Stephan,

We've been working on a few different approaches to skeletonizing large volumes including with Kimimaro, which is the most mature. You're right that stitching introduces artifacts, but our approach with  Kimimaro has been to try to either avoid particular artifacts, or fix them afterwards. Larger contexts can help reduce errors, but some issues we are dealing with are self-touches, accidental loops, and "ticks" some of which we are having better luck with than others. 

Another approach is to try skeletonizing meshes, which might allow for the algorithm to have better context.

I hope that helps!
Will

**Comment by unidesigner** [comment]
Hi Will,

Yes, this is helpful. I came across a recent CVPR2019 publication (https://danielhaehn.com/papers/matejek2019biologically.pdf) and the generated skeletons in Figure 5 look quite nice. I don't know how their implemented topological-thinning algorithm compares to e.g. Kimimaro's results. It could be nice to have "benchmark" dataset to apply and compare different skeletonization methods against each other. Perhaps some skeletal representations are more useful for certain purposes than others.

The mesh-based approach is certainly an interesting direction and I could imagine, with additional context, might produce fewer artefacts (and is computationally more efficient if meshes are available). Do you know of any usable mesh-based implementation?

Best, Stephan

**Comment by william-silversmith** [comment]
Hi Stephan,

>  I came across a recent CVPR2019 publication (https://danielhaehn.com/papers/matejek2019biologically.pdf) and the generated skeletons in Figure 5 look quite nice. I don't know how their implemented topological-thinning algorithm compares to e.g. Kimimaro's results. 

Thank you for the link Stephan, those do look nice. Topological thinning is often considered a superior approach in many applications as it preserves topology and generates "grass fire" skeletons. However, it is also somewhat slow, not easily tunable, and it preserves loops in structures which is often inappropriate for neurons.

On an engineering level, Kimimaro (the library, and to a certain degree TEASAR, the algorithm) is specifically designed for mass production of skeletons from a densely labeled in-core image. Many algorithms perform well on single binary images, but need tuning to work efficiently on thousands of labels. The out-of-core skeleton fusion complement to Kimimaro is functioning, but is less mature and not located in this repository (it may migrate here if it makes sense later on).

> Perhaps some skeletal representations are more useful for certain purposes than others.

For the problem of neuronal representation, we often like to represent neurons as trees (no loops). Sometimes wish to capture spines and other times we want to avoid them. Spine capture/avoidance can probably be performed with a thinning algorithm by post-processing, but I haven't looked into it. I'm sure there are other reasons to use different representations as well.

> It could be nice to have "benchmark" dataset to apply and compare different skeletonization methods against each other.

For evaluating and debugging in-core Kimimaro, I've been using various test datasets including SNEMI3D and cutouts from some of our larger segmentations. I've found that SNEMI3D is good for very basic validation, but it is a little small as typically skeletonization is run on mip 3 (I've been using mip 1 on SNEMI3D). It also contains only hundreds of labels, whereas a more typical production scale example is a 512x512x512 cube containing thousands of labels. It's also important to include at least one full soma in a benchmark dataset. Sometimes people make fun of me for including them in my test sets, but glia can provide indicators of whether the algorithm does something reasonable under strenuous conditions.

I've found it useful to have "small" benchmarks (512x512x100 and 512x512x512) for in-core and "large" benchmarks (2048x2048x512 or bigger) for out of core testing.

> The mesh-based approach is certainly an interesting direction and I could imagine, with additional context, might produce fewer artefacts (and is computationally more efficient if meshes are available). Do you know of any usable mesh-based implementation? 

For our internal operations, generating meshes is part of our standard dataset visualization pipeline, so often they can be assumed to be available. Other labs, especially ones working with smaller volumes, might be apt to skip that step and might prefer direct skeletonization.

I know of a mesh based skeletonization implementation that @sdorkenw and @fcollman have been experimenting with, but it currently has a number of important limitations. None of those issues are fundamental, so I expect to see progress in implementation quality over time. Sven said you can discuss with him.

Also a quick google revealed this: http://visgraph.cse.ust.hk/projects/skeleton/

Will

**Comment by fcollman** [comment]
I think everything Will said was spot on.  As he mentioned we have been experimenting with mesh based approaches as well, which could be computationally less expensive and easier to get the full neuron's context in order to side step some of the issues presented from having to do merging of voxel based approaches (if you are in the regime where you cannot fit the entire segmentation into memory of a single process).  

our code is currently located [here](https://github.com/sdorkenw/MeshParty/blob/master/meshparty/skeletonize.py)
and is essentially a graph based form of the TEASAR algorithm.  To run efficiently it required a customization of scipy which has recently been [incorporated into the master branch ](https://github.com/scipy/scipy/pull/9664)
 but has not yet been released so you must compile scipy from master source. 

The important limitations of this include.  

1. that we don't have a good way of handling disconnected meshes, so the result only works automatically and well over single connected component meshes.  We have some experimental merge tip code in the repository now, but I don't think it covers many problem areas and I don't think it's a good final direction.  If your meshes are perfect, this isn't a problem.  I'll note this is equally a problem for voxel based TEASAR, or thinning approaches. 
2. Soma handling is done assuming you have a point to locate the soma, and we don't have any automated way of setting that as the root.  
3. Self touches (i.e. the axon and dendrite of the same neuron coming into contact) are a difficult problem and many mesh generation and processing libraries love to remove duplicate vertices, we still struggle with making sure our mesh generation pipelines don't incorrectly merge meshes at self touches. 
4. Centering.  As the mesh skeleton is done across the mesh graph the nodes are by default on the surface of the mesh.    For some skeleton use cases this might not be essential, for others it might be critical.  There is a vtk based recentering approach in the repo now, that works OK on straight segments, but does not perform well on branch points.   A ray tracing based approach might improve things here but hasn't been developed.
5. Parameter tuning.  As with I think all these approaches, it's not clear that a single set of parameters will work exactly as a biologist wants them to across all types and parts of neurons.  I don't think yet fully appreciate or understand how difficult it will be to find and tune algorithm parameters to work well across a full range of data.  This is equally a caveat of any skeleton approach.

**Comment by william-silversmith** [comment]
@unidesigner If you're still interested in trying this style of skeletonization, I've merged our beta code into Igneous: https://github.com/seung-lab/igneous

**Comment by unidesigner** [comment]
@william-silversmith I'm still very much interested but other things took priority. I will get back to the skeletonization soon, also to replying to your and Forrest's detailed answers (thanks for that!). I haven't used Igneous so far, but this would be a good opportunity to do so.

**Comment by unidesigner** [comment]
I would like to run kimimaro.skeletonize on a (712, 712, 336) 'uint32' label volume, for now using just using a the suggested parameters:
```
    ...: skels = kimimaro.skeletonize( 
    ...:   labels,  
    ...:   teasar_params={ 
    ...:     'scale': 4, 
    ...:     'const': 500, # physical units 
    ...:     'pdrf_exponent': 4, 
    ...:     'pdrf_scale': 100000, 
    ...:     'soma_detection_threshold': 1100, # physical units 
    ...:     'soma_acceptance_threshold': 3500, # physical units 
    ...:     'soma_invalidation_scale': 1.0, 
    ...:     'soma_invalidation_const': 300, # physical units 
    ...:     'max_paths': 15, # default None 
    ...:   }, 
    ...:   dust_threshold=1000, 
    ...:   anisotropy=(10,10,25), # default True 
    ...:   fix_branching=True, # default True 
    ...:   fix_borders=True, # default True 
    ...:   progress=True, # default False 
    ...:   object_ids=[5649660] 
    ...: )    
```
I get the following exception:
```
    220     plane_targets = kimimaro.skeletontricks.find_border_targets(
--> 221       dt_plane, cc_plane
    222     )
    223 
ext/skeletontricks/skeletontricks.pyx in skeletontricks.find_border_targets()
ValueError: Buffer has wrong number of dimensions (expected 2, got 3)
```
Do you know what the problem may be?

**Comment by william-silversmith** [comment]
Hi Stephan! 

I just released a revised version of the code last night: 0.3.1. 0.3.0 had this bug in it, make sure you're on the latest version! If you still have this problem, try downgrading to 0.2.2 which was working well for months but doesn't have the border fix in it.

**Comment by william-silversmith** [comment]
Actually, never mind. It's that either `dt_plane` or `cc_plane` has 3 dimensions (probably a trivial third dimension) when they should just be squeezed to two. That didn't show up in my testing. Let me try an experiment...

**Comment by william-silversmith** [comment]
Another thing that could possibly cause this issue is if your connected-components-3d library is a few months old. Make sure that you have the latest versions of `edt`, `connected-components-3d`, and `fastremap`

**Comment by unidesigner** [comment]
The shapes are
```
ipdb> dt_plane.shape                                                                                 
(712, 712, 1)
ipdb> cc_plane.shape                                                                                 
(712, 712, 1)
```
Ok, I'll try to update those libraries.

**Comment by unidesigner** [comment]
Yep, this was it:
```
      Successfully uninstalled connected-components-3d-1.0.0
Successfully installed connected-components-3d-1.0.5
```
Now it runs through. Thanks! Will get back to it later.

Btw, I came across Philipp Schlegel's (@schlegelp) reimplementation of a skeletonizer. I haven't tested it yet - would be interesting to compare to kimimaro. https://github.com/schlegelp/skeletonizer

**Comment by william-silversmith** [comment]
@unidesigner version 0.4.0 now has a parallel implementation that makes it practical to skeletonize much larger volumes in a single go. You might want to give it a look. :)

**Comment by unidesigner** [comment]
The parallel implementation works nicely. It will take me a bit more time to tune parameters and inspect results etc. I will report here again.

BTW, two more links that might be of interest to people finding this thread:
https://github.com/constantinpape/skeletor
https://github.com/jni/skan/

**Comment by schlegelp** [comment]
> Btw, I came across Philipp Schlegel's (@schlegelp) reimplementation of a skeletonizer. I haven't tested it yet - would be interesting to compare to kimimaro. https://github.com/schlegelp/skeletonizer

FYI: that re-implementation is (a) only half done (mesh contraction works but not the skeletonization) and (b) needs optimisation.

**Comment by constantinpape** [comment]
> The parallel implementation works nicely. It will take me a bit more time to tune parameters and inspect results etc. I will report here again.
> 
> BTW, two more links that might be of interest to people finding this thread:
> https://github.com/constantinpape/skeletor
> https://github.com/jni/skan/

Just to give some context, my implementation is very much based on kimimaro and not mature at all.
I would rather use kimimaro myself, but unfortunately it does not fit to the software stack I am using(no conda + not compatible wit recent numpy version). Also, for my purposes it would be more helpful to expose more low level API (skeletonize a single object) and include API for the out-of-core functionality.

I am still exploring whether it would be easier to finish my reimplementation or port kimimaro to my tech stack.

**Comment by william-silversmith** [comment]
@constantinpape I think it might be possible to use this with numpy < 1.16, but you can't use the precompiled binaries for cc3d, edt, fastremap, and dijkstra3d. I also work without conda and instead use virtualenv. If that is available for your system, using Kimimaro might be possible.  

With respect to the out-of-core functionality, I was considering moving some of it into Kimimaro itself (particularly the postprocessing items in Igneous). Would that be sufficient?

**Comment by william-silversmith** [comment]
Also, I like the name of your repo. :)

**Comment by william-silversmith** [comment]
There's a relatively new feature that I failed to document in the example but is listed under performance_tips:  
   
"If you only need a few labels skeletonized, pass in object_ids to bypass processing all the others. If object_ids contains only a single label, the masking operation will run faster."  
  
I'll update the example.

**Comment by william-silversmith** [comment]
@constantinpape I added the postprocessing logic from Igneous into Kimimaro.

**Comment by constantinpape** [comment]
@william-silversmith 

> I think it might be possible to use this with numpy < 1.16, but you can't use the precompiled binaries for cc3d, edt, fastremap, and dijkstra3d. I also work without conda and instead use virtualenv. If that is available for your system, using Kimimaro might be possible.

I have kimimaro working locally now, where I rebuilt some dependencies that were not available for numpy < 1.16 myself. Longer term, I would like to have all my dependencies as conda, to make it easier for collaborators to use my software stack, but for now this set-up is fine with me.

> With respect to the out-of-core functionality, I was considering moving some of it into Kimimaro itself (particularly the postprocessing items in Igneous). Would that be sufficient?

Yes, that would be very helpful.

> Also, I like the name of your repo. :)

Thanks :).

> There's a relatively new feature that I failed to document in the example but is listed under performance_tips:
> 
> "If you only need a few labels skeletonized, pass in object_ids to bypass processing all the others. If object_ids contains only a single label, the masking operation will run faster."

That's very helpful too. For my purposes, I will probably just write a small wrapper around `trace`.

**Comment by william-silversmith** [comment]
This library and its out-of-core component in Igneous have advanced considerably, so I'll just close this issue for now. If you have a large dataset, my stock answer is "use Igneous". We even support generating sharded skeletons to avoid incurring large costs from file creation. If you have a medium-sized dataset, you can use Kimimaro alone with a large machine.

https://github.com/seung-lab/igneous
